nexus/src/db/datastore.rs

```diff
@@ -238,6 +242,173 @@ impl DataStore {
         })
     }
 
+    /// Gets allocated regions for a disk, and the datasets to which those
+    /// regions belong.
+    ///
+    /// Note that this function does not validate liveness of the Disk, so it
+    /// may be used in a context where the disk is being deleted.
+    pub async fn get_allocated_regions(
+        &self,
+        disk_id: Uuid,
+    ) -> Result<Vec<(Dataset, Region)>, Error> {
+        use db::schema::dataset::dsl as dataset_dsl;
+        use db::schema::region::dsl as region_dsl;
+        region_dsl::region
+            .filter(region_dsl::disk_id.eq(disk_id))
+            .inner_join(
+                dataset_dsl::dataset
+                    .on(region_dsl::dataset_id.eq(dataset_dsl::id)),
+            )
+            .select((Dataset::as_select(), Region::as_select()))
+            .get_results_async::<(Dataset, Region)>(self.pool())
+            .await
+            .map_err(|e| public_error_from_diesel_pool_shouldnt_fail(e))
+    }
+
+    /// Idempotently allocates enough regions to back a disk.
+    ///
+    /// Returns the allocated regions, as well as the datasets to which they
+    /// belong.
+    pub async fn region_allocate(
+        &self,
+        disk_id: Uuid,
+        params: &params::DiskCreate,
+    ) -> Result<Vec<(Dataset, Region)>, Error> {
+        use db::schema::dataset::dsl as dataset_dsl;
+        use db::schema::region::dsl as region_dsl;
+
+        // ALLOCATION POLICY
+        //
+        // NOTE: This policy can - and should! - be changed.
+        //
+        // See https://rfd.shared.oxide.computer/rfd/0205 for a more
+        // complete discussion.
+        //
+        // It is currently acting as a placeholder, showing a feasible
+        // interaction between datasets and regions.
+        //
+        // This policy allocates regions to distinct Crucible datasets,
+        // favoring datasets with the smallest existing (summed) region
+        // sizes. Basically, "pick the datasets with the smallest load first".
+        //
+        // Longer-term, we should consider:
+        // - Storage size + remaining free space
+        // - Sled placement of datasets
+        // - What sort of loads we'd like to create (even split across all disks
+        // may not be preferable, especially if maintenance is expected)
+        #[derive(Debug, thiserror::Error)]
+        enum RegionAllocateError {
+            #[error("Not enough datasets for replicated allocation: {0}")]
+            NotEnoughDatasets(usize),
+        }
+        type TxnError = TransactionError<RegionAllocateError>;
+        let params: params::DiskCreate = params.clone();
+        self.pool()
+            .transaction(move |conn| {
```

Comment:
This is the "interactive transaction" pattern we said we wanted to avoid in RFD 192.  I gather we want to prioritize urgency here?  (This is not the first place we're doing this -- I'm coming to feel like we're punting on some of the recommendations in RFD 192 and will wind up coming back to them, which is fine as long as we're aware of what we're doing.)